ManPac/Assets/Scripts/UI/LevelUICommunicator.cs

```diff
@@ -1,23 +1,244 @@
+        {
+            CharacterStates.Inactive,
+            CharacterStates.Inactive,
+            CharacterStates.Inactive,
+            CharacterStates.Inactive
+        };
+
+        _characterUiAnimators = new[]
+        {
+            UIBlinkyImage.GetComponent<Animator>(),
+            UIPinkyImage.GetComponent<Animator>(),
+            UIInkyImage.GetComponent<Animator>(),
+            UIClydeImage.GetComponent<Animator>(),
+        };
+        
+        // Binding to events
+        PelletMediator.OnPelletPickedUp.AddListener(OnPelletPickedUp);
+        Enemy.OnBehaviourStateChanged.AddListener(OnManPacStateChange);
+        CharacterSwitcher.OnCharacterActivated.AddListener(OnCharacterSwitched);
+        DeathMediator.OnDeathStateChanged.AddListener(OnCharacterDiedChanged);
+        Enemy.OnGotHitByPlayer.AddListener(OnManPacGotHit);
+        
+        // Setting up the animators
         _powerPelletAnimator = PowerPelletImage.GetComponent<Animator>();
+        _pelletCounterAnimator = UIPelletCounter.transform.parent.GetComponent<Animator>();
+        
+        // Getting the ordinary pellet count in the scene
+        _normalPelletCount = FindObjectsByType<Pellet>(FindObjectsInactive.Exclude, FindObjectsSortMode.None)
+            .Count(pellet => pellet.Type == PelletTypes.Ordinary);
+        
+        UpdatePelletCounter();
+        PrepareDivaHealth();
     }
 
-    public void OnManPacStateChange(ManPacStates newState)
+    private void OnManPacStateChange(ManPacStates newState)
     {
         _powerPelletAnimator.SetBool(PowerPelletActiveBoolean, newState == ManPacStates.Aggressive);
     }
+
+    private void OnPelletPickedUp(int score, PelletTypes pelletType)
+    {
+        if (pelletType != PelletTypes.Ordinary)
+            return;
+
+        _normalPelletCount -= 1;
+        UpdatePelletCounter();
+        _pelletCounterAnimator.SetTrigger(PelletPickupAnimationActivation);
+    }
+
+    private void OnCharacterSwitched(PlayerCharacter nowControlledCharacter)
+    {
+        int characterUiCount = _uiImageMap.Length;
+        int nowControlledIndex = (int) nowControlledCharacter.Type;
+        for (int i = 0; i < characterUiCount; i++)
+        {
+            if (_characterStates[i] == CharacterStates.Dead)
+                continue;
+
+            _characterStates[i] = CharacterStates.Inactive;
+        }
+
+        if (_characterStates[nowControlledIndex] == CharacterStates.Inactive)
+            _characterStates[nowControlledIndex] = CharacterStates.Active;
+        
+        UpdateCharacterStateUI();
+    }
+
+    private void OnCharacterDiedChanged(PlayerCharacter playerCharacter, bool isDead)
+    {
+        int diedIndex = (int) playerCharacter.Type;
+
+        _characterStates[diedIndex] = isDead ? CharacterStates.Dead : CharacterStates.Inactive;
+        UpdateCharacterStateUI();
+    }
+
+    private void OnManPacGotHit(GameObject player)
+    {
+        UpdateDivaHealth();
+    }
```

Comment:
you can remove player since you're not using it
```suggestion
    private void OnManPacGotHit()
    {
        UpdateDivaHealth();
    }
```

Reply:
Unity-event